django: insert with expression fails

Probably the [current approach for inserting](https://github.com/orijtech/spanner-orm/blob/b3f138444903c1ecd3b480b1fea2c1da64d9bcf6/spanner/dbapi/cursor.py#L139-L156) (which doesn't use SQL) doesn't support expressions.
```
======================================================================
FAIL: test_bulk_insert_expressions (bulk_create.tests.BulkCreateTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/tim/code/django/django/test/testcases.py", line 1229, in skip_wrapper
    return test_func(*args, **kwargs)
  File "/home/tim/code/django/tests/bulk_create/tests.py", line 214, in test_bulk_insert_expressions
    self.assertEqual(bbb.count(), 1)
AssertionError: 0 != 1

----------------------------------------------------------------------
(0.000) QUERY = 'INSERT INTO bulk_create_restaurant (id, name) VALUES (%s, %s), (%s, LOWER(%s))' - PARAMS = (4526302679840324305, "Sam's Shake Shack", 3661758548926185923, "Betty's Beetroot Bar"); args=(4526302679840324305, "Sam's Shake Shack", 3661758548926185923, "Betty's Beetroot Bar")
```

--- a/django_test_suite.sh
+++ b/django_test_suite.sh
@@ -7,7 +7,7 @@ set -e
 # so that we can have multiple tests running without
 # conflicting which changes and constraints. We'll always
 # cleanup the created database.
-TEST_DBNAME=${SPANNER_TEST_DB:-$(python3 -c 'import os; print("a"+os.urandom(10).hex())')}
+TEST_DBNAME=${SPANNER_TEST_DB:-$(python3 -c 'import os, time; print(chr(ord("a") + time.time_ns() % 26)+os.urandom(10).hex())')}
 TEST_DBNAME_OTHER="$TEST_DBNAME-ot"
 TEST_APPS=${DJANGO_TEST_APPS:-basic}
 INSTANCE=${SPANNER_TEST_INSTANCE:-django-tests}

--- a/spanner/dbapi/cursor.py
+++ b/spanner/dbapi/cursor.py
@@ -18,7 +18,7 @@ from .exceptions import IntegrityError, OperationalError, ProgrammingError
 from .parse_utils import (
     STMT_DDL, STMT_INSERT, STMT_NON_UPDATING, classify_stmt,
     ensure_where_clause, get_param_types, parse_insert,
-    rows_for_insert_or_update, sql_pyformat_args_to_spanner,
+    sql_pyformat_args_to_spanner,
 )
 
 _UNSET_COUNT = -1
@@ -121,29 +121,52 @@ class Cursor(object):
     def __handle_insert(self, sql, params):
         self.__commit_preceding_batch(OP_DDL)
 
-        # There are 3 variants of an INSERT statement:
-        # a) INSERT INTO <table> (columns...) VALUES (<inlined values>): no params
-        # b) INSERT INTO <table> (columns...) SELECT_STMT:               no params
-        # c) INSERT INTO <table> (columns...) VALUES (%s,...):           with params
-        parts = parse_insert(sql)
+        parts = parse_insert(sql, params)
+
+        # The split between the two styles exists because:
+        # in the common case of multiple values being passed
+        # with simple pyformat arguments,
+        #   SQL: INSERT INTO T (f1, f2) VALUES (%s, %s, %s)
+        #   Params:   [(1, 2, 3, 4, 5, 6, 7, 8, 9, 10,)]
+        # we can take advantage of a single RPC with:
+        #       transaction.insert(table, columns, values)
+        # instead of invoking:
+        #   with transaction:
+        #       for sql, params in sql_params_list:
+        #           transaction.execute_sql(sql, params, param_types)
+        # which invokes more RPCs and is more costly.
+
+        if parts.get('homogenous'):
+            # The common case of multiple values being passed in
+            # non-complex pyformat args and need to be uploaded in one RPC.
+            return self.__db_handle.in_transaction(
+                self.__do_execute_insert_homogenous,
+                parts,
+            )
+        else:
+            # All the other cases that are esoteric and need
+            #   transaction.execute_sql
+            sql_params_list = parts.get('sql_params_list')
+            return self.__db_handle.in_transaction(
+                self.__do_execute_insert_heterogenous,
+                sql_params_list,
+            )
+
+    def __do_execute_insert_heterogenous(self, transaction, sql_params_list):
+        for sql, params in sql_params_list:
+            sql, params = sql_pyformat_args_to_spanner(sql, params)
+            param_types = get_param_types(params)
+            res = transaction.execute_sql(sql, params=params, param_types=param_types)
+            # TODO: File a bug with Cloud Spanner and the Python client maintainers
+            # about a lost commit when res isn't read from.
+            _ = list(res)
+
+    def __do_execute_insert_homogenous(self, transaction, parts):
+        # Perform an insert in one shot.
+        table = parts.get('table')
         columns = parts.get('columns')
-        rows = None
-        if params:
-            # Case c)
-            rows = rows_for_insert_or_update(columns, params, parts.get('values_pyformat'))
-
-        self.__db_handle.in_transaction(
-            self.__do_execute_insert,
-            parts.get('table'),
-            columns,
-            rows,
-        )
-
-    def __do_execute_insert(self, transaction, table, columns, values):
+        values = parts.get('values')
         return transaction.insert(table, columns, values)
-
-    def __execute_insert_no_params(self, transaction, sql):
-        return transaction.execute_update(sql)
 
     def __commit_preceding_batch(self, op=None):
         last_op = self.__last_op

--- a/spanner/dbapi/parse_utils.py
+++ b/spanner/dbapi/parse_utils.py
@@ -14,11 +14,13 @@
 
 import decimal
 import re
+from functools import reduce
 
 import sqlparse
 from google.cloud import spanner_v1 as spanner
 
-from .exceptions import Error
+from .exceptions import Error, ProgrammingError
+from .parser import parse_values
 from .types import DateStr, TimestampStr
 
 STMT_DDL = 'DDL'
@@ -75,25 +77,138 @@ def strip_backticks(name):
     return name[1:-1] if has_quotes else name
 
 
-def parse_insert(insert_sql):
-    match = re_INSERT.search(insert_sql)
-    if not match:
-        return None
+def parse_insert(insert_sql, params):
+    """
+    Parse an INSERT statement an generate a list of tuples of the form:
+        [
+            (SQL, params_per_row1),
+            (SQL, params_per_row2),
+            (SQL, params_per_row3),
+            ...
+        ]
 
-    parsed = {
-        'table': strip_backticks(match.group('table_name')),
-        'columns': [
+    There are 4 variants of an INSERT statement:
+    a) INSERT INTO <table> (columns...) VALUES (<inlined values>): no params
+    b) INSERT INTO <table> (columns...) SELECT_STMT:               no params
+    c) INSERT INTO <table> (columns...) VALUES (%s,...):           with params
+    d) INSERT INTO <table> (columns...) VALUES (%s,..<EXPR>...)    with params and expressions
+
+    Thus given each of the forms, it will produce a dictionary describing
+    how to upload the contents to Cloud Spanner:
+    Case a)
+            SQL: INSERT INTO T (f1, f2) VALUES (1, 2)
+        it produces:
+            {
+                'sql_params_list': [
+                    ('INSERT INTO T (f1, f2) VALUES (1, 2)', None),
+                ],
+            }
+
+    Case b)
+            SQL: 'INSERT INTO T (s, c) SELECT st, zc FROM cus ORDER BY fn, ln',
+        it produces:
+            {
+                'sql_params_list': [
+                    ('INSERT INTO T (s, c) SELECT st, zc FROM cus ORDER BY fn, ln', None),
+                ]
+            }
+
+    Case c)
+            SQL: INSERT INTO T (f1, f2) VALUES (%s, %s), (%s, %s)
+            Params: ['a', 'b', 'c', 'd']
+        it produces:
+            {
+                'homogenous': True,
+                'table': 'T',
+                'columns': ['f1', 'f2'],
+                'values': [('a', 'b',), ('c', 'd',)],
+            }
+
+    Case d)
+            SQL: INSERT INTO T (f1, f2) VALUES (%s, LOWER(%s)), (UPPER(%s), %s)
+            Params: ['a', 'b', 'c', 'd']
+        it produces:
+            {
+                'sql_params_list': [
+                    ('INSERT INTO T (f1, f2) VALUES (%s, LOWER(%s))', ('a', 'b',))
+                    ('INSERT INTO T (f1, f2) VALUES (UPPER(%s), %s)', ('c', 'd',))
+                ],
+            }
+    """
+    match = re_INSERT.search(insert_sql)
+
+    if not match:
+        raise ProgrammingError('Could not parse an INSERT statement from %s' % insert_sql)
+
+    after_values_sql = re_VALUES_TILL_END.findall(insert_sql)
+    if not after_values_sql:
+        # Case b)
+        return {
+            'sql_params_list': [(insert_sql, None,)],
+        }
+
+    if not params:
+        # Case a) perhaps?
+        # Check if any %s exists.
+        pyformat_str_count = after_values_sql.count('%s')
+        if pyformat_str_count > 0:
+            raise ProgrammingError('no params yet there are %d "%s" tokens' % pyformat_str_count)
+
+        # Confirmed case of:
+        # SQL: INSERT INTO T (a1, a2) VALUES (1, 2)
+        # Params: None
+        return {
+            'sql_params_list': [(insert_sql, None,)],
+        }
+
+    values_str = after_values_sql[0]
+    _, values = parse_values(values_str)
+
+    if values.homogenous():
+        # Case c)
+
+        # We MUST strip backticks for the table_name and columns,
+        # only in this code path for homogenous statements.
+        table = strip_backticks(match.group('table_name'))
+
+        columns = [
             strip_backticks(mi.strip())
             for mi in match.group('columns').split(',')
-        ],
-    }
-    after_VALUES_sql = re_VALUES_TILL_END.findall(insert_sql)
-    if after_VALUES_sql:
-        values_pyformat = re_VALUES_PYFORMAT.findall(after_VALUES_sql[0])
-        if values_pyformat:
-            parsed['values_pyformat'] = values_pyformat
+        ]
+        values_pyformat = [str(arg) for arg in values.argv]
+        rows_list = rows_for_insert_or_update(columns, params, values_pyformat)
 
-    return parsed
+        return {
+            'homogenous': True,
+            'table': table,
+            'values': rows_list,
+            'columns': columns,
+        }
+
+    # Case d)
+    # insert_sql is of the form:
+    #   INSERT INTO T(c1, c2) VALUES (%s, %s), (%s, LOWER(%s))
+
+    # Sanity check:
+    #  length(all_args) == len(params)
+    args_len = reduce(lambda a, b: a+b, [len(arg) for arg in values.argv])
+    if args_len != len(params):
+        raise ProgrammingError('Invalid length: VALUES(...) len: %d != len(params): %d' % (
+            args_len, len(params)),
+        )
+
+    trim_index = insert_sql.find(values_str)
+    before_values_sql = insert_sql[:trim_index]
+
+    sql_param_tuples = []
+    for token_arg in values.argv:
+        row_sql = before_values_sql + ' VALUES%s' % token_arg
+        row_params, params = tuple(params[0:len(token_arg)]), params[len(token_arg):]
+        sql_param_tuples.append((row_sql, row_params,))
+
+    return {
+        'sql_params_list': sql_param_tuples,
+    }
 
 
 def rows_for_insert_or_update(columns, params, pyformat_args=None):
@@ -146,7 +261,7 @@ def rows_for_insert_or_update(columns, params, pyformat_args=None):
         # Columns:      (f1, f2, f3)
         # new_params:   [(1, 2, 3,), (4, 5, 6,), (7, 8, 9,)]
 
-        # Sanity check: all the pyformat_values should have the exact same length.
+        # Sanity check 1: all the pyformat_values should have the exact same length.
         first, rest = pyformat_args[0], pyformat_args[1:]
         n_stride = first.count('%s')
         for pyfmt_value in rest:
@@ -154,6 +269,19 @@ def rows_for_insert_or_update(columns, params, pyformat_args=None):
             if n_stride != n:
                 raise Error('\nlen(`%s`)=%d\n!=\nlen(`%s`)=%d' % (
                     first, n_stride, pyfmt_value, n))
+
+        # Sanity check 2: len(params) MUST be a multiple of n_stride aka len(count of %s).
+        # so that we can properly group for example:
+        #  Given pyformat args:
+        #   (%s, %s, %s)
+        #  Params:
+        #   [1, 2, 3, 4, 5, 6, 7, 8, 9]
+        # into
+        #   [(1, 2, 3), (4, 5, 6), (7, 8, 9)]
+        if (len(params) % n_stride) != 0:
+            raise ProgrammingError('Invalid length: len(params)=%d MUST be a multiple of len(pyformat_args)=%d' % (
+                len(params), n_stride),
+            )
 
     # Now chop up the strides.
     strides = []

--- a/spanner/dbapi/parser.py
+++ b/spanner/dbapi/parser.py
@@ -1,0 +1,255 @@
+# Copyright 2019 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Grammar for parsing VALUES:
+    VALUES      := `VALUES(` + ARGS + `)`
+    ARGS        := [EXPR,]*EXPR
+    EXPR        := TERMINAL / FUNC
+    TERMINAL    := `%s`
+    FUNC        := alphanum + `(` + ARGS + `)`
+    alphanum    := (a-zA-Z_)[0-9a-ZA-Z_]*
+
+thus given:
+    statement: 'VALUES (%s, %s), (%s, LOWER(UPPER(%s)))   , (%s)'
+    It'll parse:
+        VALUES
+            |- ARGS
+                |- (TERMINAL, TERMINAL)
+                |- (TERMINAL, FUNC
+                                |- FUNC
+                                    |- (TERMINAL)
+                |- (TERMINAL)
+"""
+
+from .exceptions import ProgrammingError
+
+ARGS = 'ARGS'
+EXPR = 'EXPR'
+FUNC = 'FUNC'
+TERMINAL = 'TERMINAL'
+VALUES = 'VALUES'
+
+
+class func:
+    def __init__(self, func_name, args):
+        self.name = func_name
+        self.args = args
+
+    def __str__(self):
+        return '%s%s' % (self.name, self.args)
+
+    def __repr__(self):
+        return self.__str__()
+
+    def __eq__(self, other):
+        if type(self) != type(other):
+            return False
+        if self.name != other.name:
+            return False
+        if not isinstance(other.args, type(self.args)):
+            return False
+        if len(self.args) != len(other.args):
+            return False
+        return self.args == other.args
+
+    def __len__(self):
+        return len(self.args)
+
+
+class terminal(str):
+    """
+    terminal represents the unit symbol that can be part of a SQL values clause.
+    """
+    pass
+
+
+class a_args:
+    def __init__(self, argv):
+        self.argv = argv
+
+    def __str__(self):
+        return '(' + ', '.join([str(arg) for arg in self.argv]) + ')'
+
+    def __repr__(self):
+        return self.__str__()
+
+    def has_expr(self):
+        return any([token for token in self.argv if not isinstance(token, terminal)])
+
+    def __len__(self):
+        return len(self.argv)
+
+    def __eq__(self, other):
+        if type(self) != type(other):
+            return False
+
+        s_len, o_len = len(self), len(other)
+        if s_len != o_len:
+            return False
+
+        for i, s_item in enumerate(self):
+            o_item = other[i]
+            if s_item != o_item:
+                return False
+
+        return True
+
+    def __getitem__(self, index):
+        return self.argv[index]
+
+    def homogenous(self):
+        """
+        Return True if all the arguments are pyformat
+        args and have the same number of arguments.
+        """
+        if not self.all_have_same_argc():
+            return False
+
+        for arg in self.argv:
+            if isinstance(arg, terminal):
+                continue
+            elif isinstance(arg, a_args):
+                if not arg.homogenous():
+                    return False
+            else:
+                return False
+        return True
+
+    def all_have_same_argc(self):
+        """
+        Return False if all the arguments have the same length.
+        """
+        if len(self) == 0:
+            return True
+
+        arg0_len = len(self.argv[0])
+        for arg in self.argv[1:]:
+            if len(arg) != arg0_len:
+                return False
+
+        return True
+
+
+class values(a_args):
+    def __str__(self):
+        return 'VALUES%s' % super().__str__()
+
+
+def parse_values(stmt):
+    return expect(stmt, VALUES)
+
+
+pyfmt_str = terminal('%s')
+
+
+def expect(word, token):
+    word = word.strip(' ')
+    if token == VALUES:
+        if not word.startswith('VALUES'):
+            raise ProgrammingError('VALUES: `%s` does not start with VALUES' % word)
+        word = word[len('VALUES'):].strip(' ')
+
+        all_args = []
+        while word:
+            word = word.strip(' ')
+
+            word, arg = expect(word, ARGS)
+            all_args.append(arg)
+
+            word = word.strip(' ')
+            if word == '':
+                break
+            elif word[0] != ',':
+                raise ProgrammingError('VALUES: expected `,` got %s in %s' % (word[0], word))
+            else:
+                word = word[1:]
+        return '', values(all_args)
+
+    elif token == TERMINAL:
+        word = word.strip(' ')
+        if word != '%s':
+            raise ProgrammingError('TERMINAL: `%s` is not %%s' % word)
+        return '', pyfmt_str
+
+    elif token == FUNC:
+        begins_with_letter = word and (word[0].isalpha() or word[0] == '_')
+        if not begins_with_letter:
+            raise ProgrammingError('FUNC: `%s` does not begin with `a-zA-z` nor a `_`' % word)
+
+        rest = word[1:]
+        end = 0
+        for ch in rest:
+            if ch.isalnum() or ch == '_':
+                end += 1
+            else:
+                break
+
+        func_name, rest = word[:end+1], word[end+1:].strip(' ')
+
+        word, args = expect(rest, ARGS)
+        return word, func(func_name, args)
+
+    elif token == ARGS:
+        # The form should be:
+        #   (%s)
+        #   (%s, %s...)
+        #   (FUNC, %s...)
+        #   (%s, %s...)
+        if not (word and word[0] == '('):
+            raise ProgrammingError('ARGS: supposed to begin with `(` in `%s`' % (word))
+
+        word = word[1:]
+
+        terms = []
+        while True:
+            word = word.strip(' ')
+            if not word:
+                break
+            elif word == '%s':
+                terms.append(pyfmt_str)
+                word = ''
+            elif word.startswith(')'):
+                break
+            elif not word.startswith('%s'):
+                word, parsed = expect(word, FUNC)
+                terms.append(parsed)
+            else:
+                terms.append(pyfmt_str)
+                word = word[2:].strip(' ')
+
+            if word and word[0] == ',':
+                word = word[1:]
+
+        if (not word) or word[0] != ')':
+            raise ProgrammingError('ARGS: supposed to end with `)` in `%s`' % (word))
+
+        word = word[1:]
+        return word, a_args(terms)
+
+    elif token == EXPR:
+        if word == '%s':
+            # Terminal symbol.
+            return '', (pyfmt_str)
+
+        # Otherwise we expect a function.
+        return expect(word, FUNC)
+
+    else:
+        raise ProgrammingError('Unknown token `%s`' % token)
+
+
+def as_values(values_stmt):
+    _, values = parse_values(values_stmt)
+    return values

--- a/spanner/django/features.py
+++ b/spanner/django/features.py
@@ -102,11 +102,6 @@ class DatabaseFeatures(BaseDatabaseFeatures):
         # extract() with timezone not working as expected:
         # https://github.com/orijtech/spanner-orm/issues/191
         'timezones.tests.NewDatabaseTests.test_query_datetimes',
-        # The current approach for inserting (which doesn't use SQL) doesn't
-        # support expressions: https://github.com/orijtech/spanner-orm/issues/198
-        'bulk_create.tests.BulkCreateTests.test_bulk_insert_expressions',
-        'expressions.tests.BasicExpressionsTests.test_new_object_create',
-        'expressions.tests.BasicExpressionsTests.test_new_object_save',
         # To be investigated: https://github.com/orijtech/spanner-orm/issues/135
         'admin_changelist.tests.ChangeListTests.test_multiuser_edit',
         # Implement DatabaseOperations.datetime_cast_date_sql():

--- a/tests/spanner/dbapi/test_parse_utils.py
+++ b/tests/spanner/dbapi/test_parse_utils.py
@@ -16,7 +16,7 @@ import decimal
 from unittest import TestCase
 
 from google.cloud.spanner_v1 import param_types
-from spanner.dbapi.exceptions import Error
+from spanner.dbapi.exceptions import Error, ProgrammingError
 from spanner.dbapi.parse_utils import (
     STMT_DDL, STMT_NON_UPDATING, DateStr, TimestampStr, classify_stmt,
     ensure_where_clause, escape_name, get_param_types, parse_insert,
@@ -53,43 +53,107 @@ class ParseUtilsTests(TestCase):
         cases = [
             (
                 'INSERT INTO django_migrations (app, name, applied) VALUES (%s, %s, %s)',
+                [1, 2, 3, 4, 5, 6],
                 {
-                    'table': 'django_migrations',
+                    'homogenous': True,
                     'columns': ['app', 'name', 'applied'],
-                    'values_pyformat': ['(%s, %s, %s)'],
+                    'table': 'django_migrations',
+                    'values': [(1, 2, 3,), (4, 5, 6,)],
                 },
             ),
             (
-                'INSERT INTO sales.addresses (street, city, state, zip_code)'
+                'INSERT INTO sales.addresses (street, city, state, zip_code) '
                 'SELECT street, city, state, zip_code FROM sales.customers'
                 'ORDER BY first_name, last_name',
+                None,
                 {
-                    'table': 'sales.addresses',
-                    'columns': ['street', 'city', 'state', 'zip_code'],
-                },
+                    'sql_params_list': [(
+                        'INSERT INTO sales.addresses (street, city, state, zip_code) '
+                        'SELECT street, city, state, zip_code FROM sales.customers'
+                        'ORDER BY first_name, last_name',
+                        None,
+                    )],
+                }
             ),
             (
 
-                'INSERT INTO auth_permission (name, content_type_id, codename) '
+                'INSERT INTO ap (n, ct, cn) '
                 'VALUES (%s, %s, %s), (%s, %s, %s), (%s, %s, %s),(%s,      %s, %s)',
+                (1, 2, 3, 4, 5, 6, 7, 8, 9),
                 {
-                    'table': 'auth_permission',
-                    'columns': ['name', 'content_type_id', 'codename'],
-                    'values_pyformat': ['(%s, %s, %s)', '(%s, %s, %s)', '(%s, %s, %s)', '(%s,      %s, %s)'],
+                    'homogenous': True,
+                    'columns': ['n', 'ct', 'cn'],
+                    'table': 'ap',
+                    'values': [(1, 2, 3,), (4, 5, 6,), (7, 8, 9,)],
                 },
             ),
             (
                 'INSERT INTO `no` (`yes`) VALUES (%s)',
+                (1, 4, 5),
                 {
+                    # The results MUST NOT contain any backticks.
+                    'homogenous': True,
                     'table': 'no',
                     'columns': ['yes'],
+                    'values': [(1,), (4,), (5,)],
+                },
+            ),
+            (
+                'INSERT INTO T (f1, f2) VALUES (1, 2)',
+                None,
+                {
+                    'sql_params_list': [
+                        (
+                            'INSERT INTO T (f1, f2) VALUES (1, 2)',
+                            None,
+                        ),
+                    ],
+                },
+            ),
+            (
+                'INSERT INTO `no` (`yes`, tiff) VALUES (%s, LOWER(%s)), (%s, %s), (%s, %s)',
+                (1, 'FOO', 5, 10, 11, 29),
+                {
+                    'sql_params_list': [
+                        ('INSERT INTO `no` (`yes`, tiff)  VALUES(%s, LOWER(%s))', (1, 'FOO',)),
+                        ('INSERT INTO `no` (`yes`, tiff)  VALUES(%s, %s)', (5, 10)),
+                        ('INSERT INTO `no` (`yes`, tiff)  VALUES(%s, %s)', (11, 29)),
+                    ],
                 },
             ),
         ]
-        for sql, want in cases:
+
+        for sql, params, want in cases:
             with self.subTest(sql=sql):
-                got = parse_insert(sql)
+                got = parse_insert(sql, params)
                 self.assertEqual(got, want, 'Mismatch with parse_insert of `%s`' % sql)
+
+    def test_parse_insert_invalid(self):
+        cases = [
+            (
+                'INSERT INTO django_migrations (app, name, applied) VALUES (%s, %s, %s), (%s, %s, %s)',
+                [1, 2, 3, 4, 5, 6, 7],
+                'len\\(params\\)=7 MUST be a multiple of len\\(pyformat_args\\)=3',
+            ),
+            (
+                'INSERT INTO django_migrations (app, name, applied) VALUES (%s, %s, %s), (%s, %s, LOWER(%s))',
+                [1, 2, 3, 4, 5, 6, 7],
+                'Invalid length: VALUES\\(...\\) len: 6 != len\\(params\\): 7',
+            ),
+            (
+                'INSERT INTO django_migrations (app, name, applied) VALUES (%s, %s, %s), (%s, %s, LOWER(%s)))',
+                [1, 2, 3, 4, 5, 6],
+                'VALUES: expected `,` got \\) in \\)',
+            ),
+        ]
+
+        for sql, params, wantException in cases:
+            with self.subTest(sql=sql):
+                self.assertRaisesRegex(
+                    ProgrammingError,
+                    wantException,
+                    lambda: parse_insert(sql, params),
+                )
 
     def test_rows_for_insert_or_update(self):
         cases = [

--- a/tests/spanner/dbapi/test_parser.py
+++ b/tests/spanner/dbapi/test_parser.py
@@ -1,0 +1,168 @@
+# Copyright 2019 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from unittest import TestCase
+
+from spanner.dbapi.exceptions import ProgrammingError
+from spanner.dbapi.parser import (
+    ARGS, FUNC, TERMINAL, VALUES, a_args, expect, func, pyfmt_str, values,
+)
+
+
+class ParserTests(TestCase):
+    def test_terminal(self):
+        cases = [
+            ('%s', '', pyfmt_str),
+            ('  %s', '', pyfmt_str),
+            ('  %s   ', '', pyfmt_str),
+        ]
+
+        for text, want_unconsumed, want_parsed in cases:
+            with self.subTest(text=text):
+                got_unconsumed, got_parsed = expect(text, TERMINAL)
+                self.assertEqual(got_parsed, want_parsed)
+                self.assertEqual(got_unconsumed, want_unconsumed)
+
+    def test_terminal_fail(self):
+        cases = [
+            ('', 'TERMINAL: `` is not %s'),
+            ('fdp', 'TERMINAL: `fdp` is not %s'),
+            ('%%s', 'TERMINAL: `%%s` is not %s'),
+            ('%sa', 'TERMINAL: `%sa` is not %s'),
+        ]
+
+        for text, wantException in cases:
+            with self.subTest(text=text):
+                self.assertRaisesRegex(ProgrammingError, wantException, lambda: expect(text, TERMINAL))
+
+    def test_func(self):
+        cases = [
+            ('_91())', ')', func('_91', a_args([]))),
+            ('_a()', '', func('_a', a_args([]))),
+            ('___()', '', func('___', a_args([]))),
+            ('abc()', '', func('abc', a_args([]))),
+            (
+                'AF112(%s, LOWER(%s, %s), rand(%s, %s, TAN(%s, %s)))',
+                '',
+                func('AF112', a_args([
+                    pyfmt_str,
+                    func('LOWER', a_args([pyfmt_str, pyfmt_str])),
+                    func('rand', a_args([
+                        pyfmt_str, pyfmt_str, func('TAN', a_args([pyfmt_str, pyfmt_str])),
+                    ])),
+                ])),
+            ),
+        ]
+
+        for text, want_unconsumed, want_parsed in cases:
+            with self.subTest(text=text):
+                got_unconsumed, got_parsed = expect(text, FUNC)
+                self.assertEqual(got_parsed, want_parsed)
+                self.assertEqual(got_unconsumed, want_unconsumed)
+
+    def test_func_fail(self):
+        cases = [
+            ('', 'FUNC: `` does not begin with `a-zA-z` nor a `_`'),
+            ('91', 'FUNC: `91` does not begin with `a-zA-z` nor a `_`'),
+            ('_91', 'supposed to begin with `\\(`'),
+            ('_91(', 'supposed to end with `\\)`'),
+            ('_.()', 'supposed to begin with `\\(`'),
+            ('_a.b()', 'supposed to begin with `\\(`'),
+        ]
+
+        for text, wantException in cases:
+            with self.subTest(text=text):
+                self.assertRaisesRegex(ProgrammingError, wantException, lambda: expect(text, FUNC))
+
+    def test_a_args(self):
+        cases = [
+            ('()', '', a_args([])),
+            ('(%s)', '', a_args([pyfmt_str])),
+            ('(%s,)', '', a_args([pyfmt_str])),
+            ('(%s),', ',', a_args([pyfmt_str])),
+            (
+                '(%s,%s, f1(%s, %s))', '', a_args([
+                    pyfmt_str, pyfmt_str, func('f1', a_args([pyfmt_str, pyfmt_str])),
+                ]),
+            ),
+        ]
+
+        for text, want_unconsumed, want_parsed in cases:
+            with self.subTest(text=text):
+                got_unconsumed, got_parsed = expect(text, ARGS)
+                self.assertEqual(got_parsed, want_parsed)
+                self.assertEqual(got_unconsumed, want_unconsumed)
+
+    def test_a_args_fail(self):
+        cases = [
+            ('', 'ARGS: supposed to begin with `\\(`'),
+            ('(',  'ARGS: supposed to end with `\\)`'),
+            (')',  'ARGS: supposed to begin with `\\(`'),
+            ('(%s,%s, f1(%s, %s), %s', 'ARGS: supposed to end with `\\)`'),
+        ]
+
+        for text, wantException in cases:
+            with self.subTest(text=text):
+                self.assertRaisesRegex(ProgrammingError, wantException, lambda: expect(text, ARGS))
+
+    def test_expect_values(self):
+        cases = [
+            ('VALUES ()', '', values([a_args([])])),
+            ('VALUES', '', values([])),
+            ('VALUES(%s)', '', values([a_args([pyfmt_str])])),
+            ('    VALUES    (%s)    ', '', values([a_args([pyfmt_str])])),
+            ('VALUES(%s, %s)', '', values([a_args([pyfmt_str, pyfmt_str])])),
+            (
+                'VALUES(%s, %s, LOWER(%s, %s))', '',
+                values([
+                    a_args([
+                        pyfmt_str,
+                        pyfmt_str,
+                        func('LOWER', a_args([pyfmt_str, pyfmt_str])),
+                    ]),
+                ]),
+            ),
+            (
+                'VALUES (UPPER(%s)), (%s)',
+                '',
+                values([
+                    a_args([
+                        func('UPPER', a_args([pyfmt_str])),
+                    ]),
+                    a_args([
+                        pyfmt_str,
+                    ]),
+                ]),
+            ),
+        ]
+
+        for text, want_unconsumed, want_parsed in cases:
+            with self.subTest(text=text):
+                got_unconsumed, got_parsed = expect(text, VALUES)
+                self.assertEqual(got_parsed, want_parsed)
+                self.assertEqual(got_unconsumed, want_unconsumed)
+
+    def test_expect_values_fail(self):
+        cases = [
+            ('', 'VALUES: `` does not start with VALUES'),
+            (
+                'VALUES(%s, %s, (%s, %s))',
+                'FUNC: `\\(%s, %s\\)\\)` does not begin with `a-zA-z` nor a `_`',
+            ),
+            ('VALUES(%s),,', 'ARGS: supposed to begin with `\\(` in `,`'),
+        ]
+
+        for text, wantException in cases:
+            with self.subTest(text=text):
+                self.assertRaisesRegex(ProgrammingError, wantException, lambda: expect(text, VALUES))

